demos/test_vlc_net/src/server.rs

```diff
@@ -0,0 +1,845 @@
+            // must sync update
+            let mut state = self.state.write().await;
+            let send_counts = state.metrics.send_count;
+            state.clock = state.clock.update([].iter(), node_id);
+            state.metrics.clock_update_count = state.metrics.clock_update_count + 1;
+            let mut event = P2PMsg {
+                clock: state.clock.clone(),
+                data: payload_data.clone(),
+                pub_key: None,
+                signature: None,
+            };
+
+            // Serialize and sign the event
+            let serialized_event = match bincode::options().serialize(&event) {
+                Ok(vec) => vec,
+                Err(e) => {
+                    error!("Failed to serialize event: {}", e);
+                    return None;
+                }
+            };
+
+            let signature = match self.local_key.sign(&serialized_event) {
+                Ok(sig) => sig,
+                Err(e) => {
+                    error!("Failed to sign event: {}", e);
+                    return None;
+                }
+            };
+
+            event.signature = Some(signature);
+            event.pub_key = Some(pub_key.clone());
+            let message = match bincode::options().serialize(&event) {
+                Ok(msg) => msg,
+                Err(e) => {
+                    error!("Failed to serialize signed event: {}", e);
+                    return None;
+                }
+            };
+            let message_len = message.len();
+            (event, message, message_len, send_counts)
+        };
+        Some((event, message, message_len, send_counts))
+    }
+
+    async fn event_handler(
+        self: Arc<Self>,
+        mut rx_net: mpsc::Receiver<Event>,
+        tx_async_state: mpsc::Sender<StateUpdate>,
+    ) {
+        if self.config.bootnodes.is_none() {
+            info!("bootnode don't parse events");
+            return;
+        }
+        let semaphore = Arc::new(Semaphore::new(self.config.concurrent_verify as usize));
+
+        // Event loop to handle incoming messages
+        loop {
+            let event = match rx_net.recv().await {
+                Some(ev) => ev,
+                None => break, // Receiver has been closed
+            };
+
+            match event {
+                Event::PeerConnected(peer_id) => {
+                    debug!("connected to {:?}", peer_id);
+                }
+                Event::PeerSubscribed(peer_id, topic_hash) => {
+                    debug!("Remote peer {:?} subscribed to {:?}", peer_id, topic_hash);
+                }
+                Event::Message(msg) => {
+                    // self.clone().handle_p2p_msg(msg, tx_async_state.clone()).await;
+                    let self_clone = self.clone();
+                    let tx_state_clone = tx_async_state.clone();
+                    if let Ok(_permit) = semaphore.try_acquire() {
+                        tokio::spawn(async move {
```

Comment:
Unbounded task spawning will lead to resource exhaustion, both CPU and memory

Reply:
Spawn task on each msg will lose the order of each msg received from p2p network

Reply:
- It's the bounded channel and its capacity are 1000 first, and the semaphore is used to control concurrent verification signatures, and it works well, I have checked its status by tokio-console utility.
- Message order is not considered in there, because this progress in there used the asynchronous and concurrence to handle msg and update state.

Reply:
What's the point of using a bounded channel to control capacity and semaphore together? Which one has the effect?
Also using bound channel to pipeline infinite p2p message is not a good idea, will lead to msg loss.

It makes no sense to destory the msg order here

Reply:
A bound channel will be blocked when it becomes full.  why will message loss happen? The bound channel is used to control and sync the high-pressure event flow. Are you sure the unbound channel is more suitable? I can't agree with your opinion.

You can't keep the same event order when the updating state is in an asynchronous and concurrence context. The concurrent verification signatures feature was an important demand from the beginning of the conference.